File: Algorithms/743. Network Delay Time.py
import heapq

class Solution:
    def networkDelayTime(self, times: List[List[int]], N: int, K: int) -> int:
        graph = []
        for i in range(N + 1):
            graph.append([])
            
        for g in times:
            graph[g[0]].append((g[2], g[1]))
        
        inf = 1000000000
        dist = [inf] * (N + 1)
        vis = [False] * (N + 1)
        pq = []
        
        heapq.heappush(pq, (0, K))
        dist[K] = 0
        
        while len(pq):
            u = heapq.heappop(pq)
            # print("in", u[1], vis[u[1]])
            
            if vis[u[1]]:
                continue
            
            for neighbors in graph[u[1]]:
                #print("neigh", neighbors, vis[neighbors[1]], dist[neighbors[1]], dist[u[1]] + neighbors[0])
                if not vis[neighbors[1]] and dist[neighbors[1]] > dist[u[1]] + neighbors[0]:
                    dist[neighbors[1]] = dist[u[1]] + neighbors[0]
                    heapq.heappush(pq, (dist[neighbors[1]], neighbors[1]))
            
            vis[u[1]] = True
        
        for i in range(1, N + 1, 1):
            if vis[i] == False:
                return -1
        
        return max(dist[1:])

File: Daily Challenges/2022/May/743. Network Delay Time.cpp
struct node {
    int at, cost;
    node(){}
    node(int at, int cost)
    {
        this->at = at;
        this->cost = cost;
    }
};

struct edge {
    int v, w;
    edge(){}
    edge(int v, int w)
    {
        this->v = v;
        this->w = w;
    }
};

class comparator
{
public:
    bool operator() (node a, node b)
    {
       return a.cost > b.cost;
    }
};

class Solution {
public:
    int dijkstra(int n, int k, vector<vector<edge>> adj)
    {
        vector<bool> vis(n + 1, false);
        vector<int> dist(n + 1, INT_MAX);
        
        priority_queue<node, vector<node>, comparator> pq;
        
        dist[k] = 0;
        pq.push(node(k, 0));
        
        while(!pq.empty())
        {
            node u = pq.top();
            pq.pop();
            
            if(vis[u.at])
                continue;
            
            for(edge e : adj[u.at])
            {
                if(!vis[e.v] && dist[e.v] > dist[u.at] + e.w)
                {
                    dist[e.v] = dist[u.at] + e.w;
                    pq.push(node(e.v, dist[e.v]));
                }
            }
            
            vis[u.at] = true;
        }
        
        int mx = 0;
        for(int i = 1; i <= n; i++)
        {
            if(!vis[i]) return -1;
            
            if(i != k) {
                mx = max(mx, dist[i]);
            }
        }
        
        return mx;
    }
    
    
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<vector<edge>> adj(n + 1);
        for(int i = 0; i < times.size(); i++)
            adj[times[i][0]].push_back(edge(times[i][1], times[i][2]));
        
        return dijkstra(n, k, adj);
    }
};
